pulp_container/app/downloaders.py

```diff
@@ -20,7 +28,41 @@
 )
 
 
-class RegistryAuthHttpDownloader(HttpDownloader):
+class ValidateResourceSizeMixin:
+    async def validate_resource_size(self, response, request_method=None):
+        """
+        Verify if the constrained resources are not exceeding the maximum size allowed.
+        """
+        if request_method == "head":
+            return
+
+        content_type = response.content_type
+        max_resource_size = 0
+        is_cosign_tag = fnmatch.fnmatch(response.url.name, "sha256-*.sig")
+
+        if isinstance(self, NoAuthSignatureDownloader) or is_cosign_tag:
+            max_resource_size = SIGNATURE_PAYLOAD_MAX_SIZE
+            content_type = "Signature"
+        elif content_type in MANIFEST_MEDIA_TYPES.IMAGE + MANIFEST_MEDIA_TYPES.LIST:
+            max_resource_size = MANIFEST_PAYLOAD_MAX_SIZE
+            content_type = "Manifest"
+        else:
+            return
+
+        total_size = 0
+        buffer = b""
+        async for chunk in response.content.iter_chunked(MEGABYTE):
```

Comment:
I'm not familiar with the codebase, but would it make sense to send HTTP 413 Request Entity Too Large **before** you start reading by inspecting the `Content-Length` header (which I already saw was done elsewhere).

Reply:
hum... it seems like a good idea, but my concern is that it is easy to modify the header and bypass the verification, so I'm not sure if we can trust only in the `Content-Length` value, what do you think?

Reply:
If you really want to be safe, you can apply both. First check the header isn't bigger than the maximum size and then make you read at most the `Content-Length` size. Though I think `Content-Length` is an optional header so if it's not present, you should keep the current code anyway.

I'd recommend reading aiohttp's documentation on what you can (and can't) assume.